File: 3. Longest Substring Without Repeating Characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) < 2:
            return len(s)
        
        maxlen = 0
        
        i = 0
        j = 0
        seen = {}
        while j < len(s):
            maxlen = max(maxlen, j-i)
            if s[j] in seen:
                i = max(i, seen[s[j]]+1)
            
            seen[s[j]] = j
            
            j += 1
            
        return max(maxlen, j-i)

File: 3. Longest Substring Without Repeating Characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() <= 1) {
            return s.length();
        }
        
        int len = 0;
        int j = 0;
        String key;
        HashMap<String, Integer> index = new HashMap();
        for (int i = 0; i < s.length(); i++) {
            key = Character.toString(s.charAt(i));
            if (index.containsKey(key) == true) {
                j = Math.max(j, index.get(key) + 1);
            }
            index.put(key, i);
            if (len < (i - j + 1)) {
                len = i - j + 1;
            }
        }
        return len;
    }
}
